### PR TITLE
Improve "need `this`" error

### PR DESCRIPTION
**Pull request by dkorpel**
https://forum.dlang.org/post/bfoyxbqcevunbicxjvny@forum.dlang.org

### CODE REVIEW

**Comment by SixthDot** [comment]
The new diagnostics are better. The only problem I see is that the PR is not linked to an issue so it wont appear in the next changelog. Maybe add a changelog entry or assuming everybody hated the previous messages, create an issue such as "'need this' error message are not really useful".

**Comment by dkorpel** [comment]
Error message improvements don't need a changelog entry

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Please post this to the forum thread so that our users can see that we are fixing things that they bring up.

**Comment by Geod24** [comment]
This is so much better, thank you

**Comment by maxhaton** [comment]
> Error message improvements don't need a changelog entry

People like reading changelogs, don't be afraid. It's the same psychology as loading bars making GUI users happier.

**Comment by maxhaton** [comment]
Shouldn't this have gone to stable

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -2619,6 +2619,14 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         result = ErrorExp.get();
     }
 
+    private void needThisError(Loc loc, FuncDeclaration f)
+    {
+        auto t = f.isThis();
+        assert(t);
+        .error(loc, "calling non-static function `%s` requires an instance of type `%s`", f.toChars(), t.toChars());
```
Comment (WalterBright) [inline compiler/src/dmd/expressionsem.d:2626]: Calling it a `member` function is more precise.

File: compiler/src/dmd/expression.d [hunk 2]
```diff
@@ -1620,7 +1620,9 @@ extern (C++) abstract class Expression : ASTNode
             {
                 //printf("checkRightThis sc.intypeof = %d, ad = %p, func = %p, fdthis = %p\n",
                 //        sc.intypeof, sc.getStructClassScope(), func, fdthis);
-                error("need `this` for `%s` of type `%s`", ve.var.toChars(), ve.var.type.toChars());
+                auto t = ve.var.isThis();
+                assert(t);
+                error("accessing non-static variable `%s` requires an instance of `%s`", ve.var.toChars(), t.toChars());
```
Comment (WalterBright) [inline compiler/src/dmd/expression.d:1625]: Pedantically, struct class member variables are a `field` not a `variable`.

File: compiler/src/dmd/expression.d [hunk 3]
```diff
@@ -3798,7 +3800,11 @@ extern (C++) final class SymOffExp : SymbolExp
             // FIXME: This error report will never be handled anyone.
             // It should be done before the SymOffExp construction.
             if (v.needThis())
-                .error(loc, "need `this` for address of `%s`", v.toChars());
+            {
+                auto t = v.isThis();
+                assert(t);
+                .error(loc, "taking the address of non-static variable `%s` requires an instance of `%s`", v.toChars(), t.toChars());
```
Comment (WalterBright) [inline compiler/src/dmd/expression.d:3806]: `member`

**Pull request merged**